### PR TITLE
Fix warnings for Flutter 3.0 - `Warning: Operand of null-aware operation '!' has type 'SchedulerBinding' which excludes null.`

### LINKED ISSUE #771: Fix: Operand of null-aware operation '!' has type 'SchedulerBinding' which excludes null
if anyone facing the warning with charts_flutter package after the `flutter 3.0` update , you can simply remove null check `'!'` on respective code lines, and rebuild the app. @google-admin please fix the issue on your side.

Screenshot 2022-05-23 at 1 12 39 PM

### PR DESCRIPTION
**Pull request by fzyzcjy**
Detailed warnings:

```
    ../../../../.pub-cache/git/charts-30477090290b348ed3101bc13017aae465f59017/charts_flutter/lib/src/chart_container.dart:249:27: Warning: Operand of null-aware operation '!' has type 'SchedulerBinding' which excludes null.
     - 'SchedulerBinding' is from 'package:flutter/src/scheduler/binding.dart' ('../../../../fvm/versions/3.0.0/packages/flutter/lib/src/scheduler/binding.dart').
        if (!SchedulerBinding.instance!.hasScheduledFrame) {
                              ^
    ../../../../.pub-cache/git/charts-30477090290b348ed3101bc13017aae465f59017/charts_flutter/lib/src/chart_container.dart:250:24: Warning: Operand of null-aware operation '!' has type 'SchedulerBinding' which excludes null.
     - 'SchedulerBinding' is from 'package:flutter/src/scheduler/binding.dart' ('../../../../fvm/versions/3.0.0/packages/flutter/lib/src/scheduler/binding.dart').
          SchedulerBinding.instance!.scheduleFrame();
                           ^
    ../../../../.pub-cache/git/charts-30477090290b348ed3101bc13017aae465f59017/charts_flutter/lib/src/chart_container.dart:253:22: Warning: Operand of null-aware operation '!' has type 'SchedulerBinding' which excludes null.
     - 'SchedulerBinding' is from 'package:flutter/src/scheduler/binding.dart' ('../../../../fvm/versions/3.0.0/packages/flutter/lib/src/scheduler/binding.dart').
        SchedulerBinding.instance!.addPostFrameCallback(startAnimationController);
                         ^
    ../../../../.pub-cache/git/charts-30477090290b348ed3101bc13017aae465f59017/charts_flutter/lib/src/chart_container.dart:276:22: Warning: Operand of null-aware operation '!' has type 'SchedulerBinding' which excludes null.
     - 'SchedulerBinding' is from 'package:flutter/src/scheduler/binding.dart' ('../../../../fvm/versions/3.0.0/packages/flutter/lib/src/scheduler/binding.dart').
        SchedulerBinding.instance!.addPostFrameCallback(doRebuild);
```

fix #771

### CODE REVIEW

**Comment by viralkachhadiya** [comment]
@fzyzcjy  one more file need to fix https://github.com/google/charts/blob/master/charts_flutter/example/lib/behaviors/slider.dart
line no.122

**Comment by fzyzcjy** [comment]
done

**Comment by viralkachhadiya** [comment]
must be specify limit new flutter version 3.0.0 in pubspec.yml otherwise those who are not using latest version of 3.0.0, they all are getting error.

**Comment by fzyzcjy** [comment]
done

**Comment by rickcasson** [comment]
To maintain backwards compatibility while still removing the warnings, I would suggest you use an ambiguate method as suggested in the [Flutter 3.0 Release Notes](https://docs.flutter.dev/development/tools/sdk/release-notes/release-notes-3.0.0#your-code)

File: charts_flutter/pubspec.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@ homepage: https://github.com/google/charts
 
 environment:
   sdk: '>=2.12.0 <3.0.0'
-  flutter: '>=2.5.0'
+  flutter: '>=3.0.0'
```
Comment (rickcasson) [inline charts_flutter/pubspec.yaml:8]: Can leave as   `flutter: '>=2.5.0'`

File: charts_flutter/lib/src/chart_container.dart [hunk 2]
```diff
@@ -229,7 +229,7 @@ class ChartContainerRenderObject<D> extends RenderCustomPaint
     // This is needed to request rebuild after the legend has been added in the
     // post process phase of the chart, which happens during the chart widget's
     // build cycle.
-    SchedulerBinding.instance!.addPostFrameCallback(doRebuild);
+    SchedulerBinding.instance.addPostFrameCallback(doRebuild);
```
Comment (rickcasson) [inline charts_flutter/lib/src/chart_container.dart:232]: replace with:
`_ambiguate(SchedulerBinding.instance)!.addPostFrameCallback(doRebuild);`

File: charts_flutter/lib/src/chart_container.dart [hunk 3]
```diff
@@ -202,11 +202,11 @@ class ChartContainerRenderObject<D> extends RenderCustomPaint
 
     // Sometimes chart behaviors try to draw the chart outside of a Flutter draw
     // cycle. Schedule a frame manually to handle these cases.
-    if (!SchedulerBinding.instance!.hasScheduledFrame) {
-      SchedulerBinding.instance!.scheduleFrame();
+    if (!SchedulerBinding.instance.hasScheduledFrame) {
+      SchedulerBinding.instance.scheduleFrame();
```
Comment (rickcasson) [inline charts_flutter/lib/src/chart_container.dart:206]: replace with:
```
if (!_ambiguate(SchedulerBinding.instance)!.hasScheduledFrame) {
  _ambiguate(SchedulerBinding.instance)!.scheduleFrame();
```

File: charts_flutter/example/lib/behaviors/slider.dart [hunk 4]
```diff
@@ -119,7 +119,7 @@ class _SliderCallbackState extends State<SliderLine> {
       });
     }
 
-    SchedulerBinding.instance!.addPostFrameCallback(rebuild);
+    SchedulerBinding.instance.addPostFrameCallback(rebuild);
```
Comment (rickcasson) [inline charts_flutter/example/lib/behaviors/slider.dart:122]: replace with:
`_ambiguate(SchedulerBinding.instance)!.addPostFrameCallback(rebuild);`

File: charts_flutter/lib/src/chart_container.dart [hunk 5]
```diff
@@ -229,7 +229,7 @@ class ChartContainerRenderObject<D> extends RenderCustomPaint
     // This is needed to request rebuild after the legend has been added in the
     // post process phase of the chart, which happens during the chart widget's
     // build cycle.
-    SchedulerBinding.instance!.addPostFrameCallback(doRebuild);
+    SchedulerBinding.instance.addPostFrameCallback(doRebuild);
   }
```
Comment (rickcasson) [inline charts_flutter/lib/src/chart_container.dart:234]: add:
`T? _ambiguate<T>(T? value) => value;`

File: charts_flutter/lib/src/chart_container.dart [hunk 6]
```diff
@@ -202,11 +202,11 @@ class ChartContainerRenderObject<D> extends RenderCustomPaint
...
     // Sometimes chart behaviors try to draw the chart outside of a Flutter draw
     // cycle. Schedule a frame manually to handle these cases.
-    if (!SchedulerBinding.instance!.hasScheduledFrame) {
-      SchedulerBinding.instance!.scheduleFrame();
+    if (!SchedulerBinding.instance.hasScheduledFrame) {
+      SchedulerBinding.instance.scheduleFrame();
     }
 
-    SchedulerBinding.instance!.addPostFrameCallback(startAnimationController);
+    SchedulerBinding.instance.addPostFrameCallback(startAnimationController);
```
Comment (rickcasson) [inline charts_flutter/lib/src/chart_container.dart:209]: replace with:
`_ambiguate(SchedulerBinding.instance)!.addPostFrameCallback(startAnimationController);`

File: charts_flutter/example/lib/behaviors/slider.dart [hunk 7]
```diff
@@ -119,7 +119,7 @@ class _SliderCallbackState extends State<SliderLine> {
...
   }
```
Comment (rickcasson) [inline charts_flutter/example/lib/behaviors/slider.dart:124]: add:
`T? _ambiguate<T>(T? value) => value;`

**Review by rickcasson — COMMENTED** [review]
To maintain compatibility while removing warnings, I would recommend using the approach suggested in the Flutter 3.0 release notes: https://docs.flutter.dev/development/tools/sdk/release-notes/release-notes-3.0.0#your-code

**Comment by viralkachhadiya** [comment]
@rickcasson can you close this pr? Bcz @fzyzcjy might be not available and assign me .

**Comment by fzyzcjy** [comment]
Oh I forgot it. Let me update it.

**Comment by fzyzcjy** [comment]
done

**Comment by ilovelinux** [comment]
Fix #771

**Comment by fzyzcjy** [comment]
Any updates...?

**Comment by lohanbodevan** [comment]
Hey @rickcasson, any idea when it would be released? Thanks ;)

**Comment by rickcasson** [comment]
@lohanbodevan @fzyzcjy, sorry, no idea. I'm not a maintainer of this repo, and [based on the readme](https://github.com/google/charts#development), it doesn't sound like they aren't accepting external contributions anyway...

**Comment by blong-sofi** [comment]
@google-admin , @google-ospo-team , gentle bump

**Comment by marceloamp2** [comment]
Same issue

**Review by robinpdev — COMMENTED** [review]
looks awesome, comments are clear. Would be nice if this got approved

**Comment by gfb-47** [comment]
Would be nice if this got approved

**Review by a15056478 — COMMENTED** [review]
Looks good

**Review by filippo-veneri — COMMENTED** [review]
PR is OK

**Comment by tp-coderbabu** [comment]
@nshahan can you please look into it? Thank you.

**Comment by jatinArhamLabs** [comment]
@rickcasson when this branch gets merged

**Comment by DanOxlade** [comment]
+1 for this, this is the last package my app needs to be updated for Flutter 3+

**Comment by ilovelinux** [comment]
@DanOxlade you can already upgrade to Flutter 3+. The library does its job, the only annoying thing is a little warning (not error, see PR description) you will get every time you will build an app.

**Comment by gfb-47** [comment]
pr is fine, what is the issue?

**Comment by stuartmorgan-g** [comment]
This project is no longer maintained, and cannot accept any PRs. Community members interested in collaborating on a community-maintained fork can coordinate in [this issue](https://github.com/google/charts/issues/798).

**Pull request closed without merging**